File: src/main/java/me/itzisonn_/meazy/instruction/InstructionsSet.java
package me.itzisonn_.meazy.instruction;

import lombok.Getter;
import me.itzisonn_.meazy.instruction.array.NewReferenceArrayInstruction;
import me.itzisonn_.meazy.instruction.array.StoreReferenceIntoArrayInstruction;
import me.itzisonn_.meazy.instruction.label.GotoLabelIfComparisonTrueInstruction;
import me.itzisonn_.meazy.instruction.label.GotoLabelIfComparisonTrueInstruction.ComparisonOperation;
import me.itzisonn_.meazy.instruction.label.GotoLabelIfEqualsZeroInstruction;
import me.itzisonn_.meazy.instruction.field.GetFieldInstruction;
import me.itzisonn_.meazy.instruction.field.StoreFieldInstruction;
import me.itzisonn_.meazy.instruction.field.WithFieldInstruction;
import me.itzisonn_.meazy.instruction.local.SetLocalNameInstruction;
import me.itzisonn_.meazy.instruction.method.*;
import me.itzisonn_.meazy.instruction.label.BindLabelInstruction;
import me.itzisonn_.meazy.instruction.label.GotoLabelIfNonNullInstruction;
import me.itzisonn_.meazy.instruction.label.GotoLabelInstruction;
import me.itzisonn_.meazy.instruction.label.InitLabelInstruction;
import me.itzisonn_.meazy.instruction.stack.LoadConstantInstruction;
import me.itzisonn_.meazy.instruction.local.GetLocalInstruction;
import me.itzisonn_.meazy.instruction.stack.LoadThisReferenceInstruction;
import me.itzisonn_.meazy.instruction.local.StoreLocalInstruction;
import me.itzisonn_.meazy.instruction.misc.CheckCastInstruction;
import me.itzisonn_.meazy.instruction.misc.InstanceOfInstruction;
import me.itzisonn_.meazy.instruction.misc.ReturnInstruction;
import me.itzisonn_.meazy.instruction.misc.WithClassInstruction;
import me.itzisonn_.meazy.instruction.number.*;
import me.itzisonn_.meazy.instruction.number.ArithmeticOperationInstruction.ArithmeticOperation;
import me.itzisonn_.meazy.instruction.number.LogicalOperationInstruction.LogicalOperation;
import me.itzisonn_.meazy.instruction.stack.DuplicateInstruction;
import me.itzisonn_.meazy.instruction.stack.PopInstruction;
import me.itzisonn_.meazy.instruction.method.InvokeMethodInstruction.InvokeType;
import org.jspecify.annotations.NullMarked;
import org.jspecify.annotations.Nullable;

import java.lang.classfile.attribute.InnerClassesAttribute;
import java.lang.constant.*;
import java.lang.reflect.AccessFlag;
import java.util.*;
import java.util.function.Consumer;

@NullMarked
public class InstructionsSet {
    private final List<Instruction> instructions = new ArrayList<>();
    @Getter
    private final BytecodeBuilders bytecodeBuilders;

    public InstructionsSet(BytecodeBuilders bytecodeBuilders) {
        this.bytecodeBuilders = bytecodeBuilders;
    }

    public List<Instruction> getInstructions() {
        return new ArrayList<>(instructions);
    }

    public void with(Instruction instruction) {
        instructions.add(instruction);
    }



    public void withField(String id, ClassDesc type, Set<AccessFlag> flags) {
        with(new WithFieldInstruction(id, type, toIntFlags(flags)));
    }

    public void getField(ClassDesc owner, String id, ClassDesc type) {
        with(new GetFieldInstruction(owner, id, type, false));
    }

    public void getStaticField(ClassDesc owner, String id, ClassDesc type) {
        with(new GetFieldInstruction(owner, id, type, true));
    }

    public void storeField(ClassDesc owner, String id, ClassDesc type) {
        with(new StoreFieldInstruction(owner, id, type, false));
    }

    public void storeStaticField(ClassDesc owner, String id, ClassDesc type) {
        with(new StoreFieldInstruction(owner, id, type, true));
    }



    public void withMethod(String id, MethodTypeDesc methodTypeDesc, Set<AccessFlag> flags, Consumer<InstructionsSet> bodyInstructions) {
        with(new WithMethodInstruction(id, methodTypeDesc, toIntFlags(flags), bodyInstructions));
    }

    public void withConstructor(MethodTypeDesc methodTypeDesc, Set<AccessFlag> flags, Consumer<InstructionsSet> bodyInstructions) {
        with(new WithConstructorInstruction(methodTypeDesc, toIntFlags(flags), bodyInstructions));
    }

    public void withClass(ClassDesc classDesc, @Nullable ClassDesc superClass, Set<ClassDesc> interfaceClasses, Set<AccessFlag> flags, List<InnerClassesAttribute> attributes, Consumer<InstructionsSet> classInstructions) {
        with(new WithClassInstruction(classDesc, superClass, interfaceClasses, toIntFlags(flags), attributes, classInstructions));
    }



    public void invokeMethod(ClassDesc owner, String id, MethodTypeDesc methodTypeDesc, Consumer<InstructionsSet> argsInstructions, InvokeType invokeType) {
        with(new InvokeMethodInstruction(owner, id, methodTypeDesc, argsInstructions, invokeType));
    }

    public void invokeDynamicMethod(DirectMethodHandleDesc bootstrapMethod, String id, MethodTypeDesc methodTypeDesc, ConstantDesc... args) {
        with(new InvokeDynamicMethodInstruction(bootstrapMethod, id, methodTypeDesc, List.of(args)));
    }

    public void invokeConstructor(ClassDesc owner, MethodTypeDesc constructorTypeDesc, Consumer<InstructionsSet> argsInstructions) {
        with(new InvokeConstructorInstruction(owner, constructorTypeDesc, argsInstructions, false));
    }

    public void invokeSuperClass(ClassDesc owner, MethodTypeDesc constructorTypeDesc, Consumer<InstructionsSet> argsInstructions) {
        with(new InvokeConstructorInstruction(owner, constructorTypeDesc, argsInstructions, true));
    }



    public void loadConstant(@Nullable ConstantDesc constant) {
        with(new LoadConstantInstruction(constant));
    }

    public void loadConstant(boolean constant) {
        loadConstant(constant ? 1 : 0);
    }

    public void loadNull() {
        loadConstant(null);
    }

    public void loadThisReference() {
        with(new LoadThisReferenceInstruction());
    }



    public void getLocal(ClassDesc type, int slot) {
        with(new GetLocalInstruction(type, slot));
    }

    public void storeLocal(ClassDesc type, int slot) {
        with(new StoreLocalInstruction(type, slot));
    }

    public void setLocalName(int slot, String id, ClassDesc type, UUID startLabelUuid, UUID endLabelUuid) {
        with(new SetLocalNameInstruction(slot, id, type, startLabelUuid, endLabelUuid));
    }



    public void newReferenceArray(ClassDesc type) {
        with(new NewReferenceArrayInstruction(type));
    }

    public void storeReferenceIntoArray() {
        with(new StoreReferenceIntoArrayInstruction());
    }



    public void negateNumber(NumberType type) {
        with(new NegateNumberInstruction(type));
    }

    public void negateNumber(ClassDesc classDesc) {
        negateNumber(getNullSafeNumberType(classDesc));
    }

    public void convertToNumberType(NumberType from, NumberType to) {
        with(new ConvertToNumberTypeInstruction(from, to));
    }

    public void convertToNumberType(ClassDesc from, ClassDesc to) {
        convertToNumberType(getNullSafeNumberType(from), getNullSafeNumberType(to));
    }

    public void convertToNumberType(NumberType from, ClassDesc to) {
        convertToNumberType(from, getNullSafeNumberType(to));
    }

    public void convertToNumberType(ClassDesc from, NumberType to) {
        convertToNumberType(getNullSafeNumberType(from), to);
    }

    public void convertToBooleanType(boolean isFromBoxed, boolean isToBoxed) {
        with(new ConvertToBooleanTypeInstruction(isFromBoxed, isToBoxed));
    }

    public void arithmeticOperation(NumberType type, ArithmeticOperation operation) {
        with(new ArithmeticOperationInstruction(type, operation));
    }

    public void arithmeticOperation(ClassDesc type, ArithmeticOperation operation) {
        arithmeticOperation(getNullSafeNumberType(type), operation);
    }

    public void logicalOperation(LogicalOperation operation) {
        with(new LogicalOperationInstruction(operation));
    }



    public void instanceOf(ClassDesc target) {
        with(new InstanceOfInstruction(target));
    }

    public void checkCast(ClassDesc type) {
        with(new CheckCastInstruction(type));
    }

    public void returnValue(@Nullable ClassDesc classDesc) {
        with(new ReturnInstruction(classDesc));
    }

    public void returnVoid() {
        returnValue(null);
    }



    public void duplicate() {
        with(new DuplicateInstruction());
    }

    public void pop() {
        with(new PopInstruction());
    }



    public UUID createLabel() {
        UUID uuid = UUID.randomUUID();
        bytecodeBuilders.addLabel(uuid);
        return uuid;
    }

    public void initLabel(UUID uuid) {
        with(new InitLabelInstruction(uuid));
    }

    public UUID createAndInitLabel() {
        UUID uuid = createLabel();
        initLabel(uuid);
        return uuid;
    }

    public void bindLabel(UUID uuid) {
        with(new BindLabelInstruction(uuid));
    }

    public void gotoLabel(UUID uuid) {
        with(new GotoLabelInstruction(uuid));
    }



    public void gotoLabelIfNonNull(UUID uuid) {
        with(new GotoLabelIfNonNullInstruction(uuid));
    }

    public void gotoLabelIfEqualsZero(UUID uuid) {
        with(new GotoLabelIfEqualsZeroInstruction(uuid));
    }

    public void gotoLabelIfComparisonTrue(NumberType type, ComparisonOperation operation, UUID uuid) {
        with(new GotoLabelIfComparisonTrueInstruction(type, operation, uuid));
    }

    public void gotoLabelIfComparisonTrue(ClassDesc type, ComparisonOperation operation, UUID uuid) {
        gotoLabelIfComparisonTrue(getNullSafeNumberType(type), operation, uuid);
    }



    private NumberType getNullSafeNumberType(ClassDesc type) {
        NumberType numberType = NumberType.valueOf(type);
        if (numberType == null) throw new IllegalArgumentException("There's no NumberType corresponding to ClassDesc " +  type);
        return numberType;
    }

    private static int toIntFlags(Collection<AccessFlag> accessFlags) {
        return accessFlags.stream().map(AccessFlag::mask).reduce((i1, i2) -> i1 | i2).orElse(0);
    }

    private static int toIntFlags(AccessFlag... accessFlags) {
        return toIntFlags(Arrays.asList(accessFlags));
    }
}

File: src/main/kotlin/me/itzisonn_/meazy/instruction/InstructionsSet.kt
package me.itzisonn_.meazy.instruction

import me.itzisonn_.meazy.instruction.array.NewReferenceArrayInstruction
import me.itzisonn_.meazy.instruction.array.StoreReferenceIntoArrayInstruction
import me.itzisonn_.meazy.instruction.field.GetFieldInstruction
import me.itzisonn_.meazy.instruction.field.StoreFieldInstruction
import me.itzisonn_.meazy.instruction.field.WithFieldInstruction
import me.itzisonn_.meazy.instruction.label.*
import me.itzisonn_.meazy.instruction.label.GotoLabelIfComparisonTrueInstruction.ComparisonOperation
import me.itzisonn_.meazy.instruction.local.GetLocalInstruction
import me.itzisonn_.meazy.instruction.local.SetLocalNameInstruction
import me.itzisonn_.meazy.instruction.local.StoreLocalInstruction
import me.itzisonn_.meazy.instruction.method.*
import me.itzisonn_.meazy.instruction.method.InvokeMethodInstruction.InvokeType
import me.itzisonn_.meazy.instruction.misc.CastInstruction
import me.itzisonn_.meazy.instruction.misc.InstanceOfInstruction
import me.itzisonn_.meazy.instruction.misc.ReturnInstruction
import me.itzisonn_.meazy.instruction.misc.WithClassInstruction
import me.itzisonn_.meazy.instruction.number.*
import me.itzisonn_.meazy.instruction.number.ArithmeticOperationInstruction.ArithmeticOperation
import me.itzisonn_.meazy.instruction.number.LogicalOperationInstruction.LogicalOperation
import me.itzisonn_.meazy.instruction.stack.DuplicateInstruction
import me.itzisonn_.meazy.instruction.stack.LoadConstantInstruction
import me.itzisonn_.meazy.instruction.stack.LoadThisReferenceInstruction
import me.itzisonn_.meazy.instruction.stack.PopInstruction
import java.lang.classfile.attribute.InnerClassesAttribute
import java.lang.constant.ClassDesc
import java.lang.constant.ConstantDesc
import java.lang.constant.DirectMethodHandleDesc
import java.lang.constant.MethodTypeDesc
import java.lang.reflect.AccessFlag
import kotlin.uuid.Uuid

class InstructionsSet(private val bytecodeBuilders: BytecodeBuilders) {
    private val _instructions: MutableList<Instruction> = mutableListOf()
    val instructions get() = _instructions.toList()

    fun with(instruction: Instruction) {
        _instructions.add(instruction)
    }



    fun withField(id: String, type: ClassDesc, flags: Set<AccessFlag>) {
        with(WithFieldInstruction(id, type, toIntFlags(flags)))
    }

    fun getField(owner: ClassDesc, id: String, type: ClassDesc) {
        with(GetFieldInstruction(owner, id, type, false))
    }

    fun getStaticField(owner: ClassDesc, id: String, type: ClassDesc) {
        with(GetFieldInstruction(owner, id, type, true))
    }

    fun storeField(owner: ClassDesc, id: String, type: ClassDesc) {
        with(StoreFieldInstruction(owner, id, type, false))
    }

    fun storeStaticField(owner: ClassDesc, id: String, type: ClassDesc) {
        with(StoreFieldInstruction(owner, id, type, true))
    }



    fun withMethod(
        id: String, methodTypeDesc: MethodTypeDesc,
        flags: Set<AccessFlag>, bodyInstructions: InstructionsSet.() -> Unit
    ) {
        with(WithMethodInstruction(
            id, methodTypeDesc, toIntFlags(flags), bodyInstructions
        ))
    }

    fun withConstructor(methodTypeDesc: MethodTypeDesc, flags: Set<AccessFlag>, bodyInstructions: InstructionsSet.() -> Unit) {
        with(WithConstructorInstruction(
            methodTypeDesc, toIntFlags(flags), bodyInstructions
        ))
    }

    fun withClass(
        classDesc: ClassDesc, superClass: ClassDesc?, interfaceClasses: Set<ClassDesc>,
        flags: Set<AccessFlag>, attributes: List<InnerClassesAttribute>, classInstructions: InstructionsSet.() -> Unit
    ) {
        with(WithClassInstruction(
            classDesc, superClass, interfaceClasses,
            toIntFlags(flags), attributes, classInstructions
        ))
    }



    fun invokeMethod(
        owner: ClassDesc, id: String, methodTypeDesc: MethodTypeDesc,
        invokeType: InvokeType, argsInstructions: InstructionsSet.() -> Unit = {}
    ) {
        with(InvokeMethodInstruction(
            owner, id, methodTypeDesc, argsInstructions, invokeType
        ))
    }

    fun invokeDynamicMethod(
        bootstrapMethod: DirectMethodHandleDesc, id: String,
        methodTypeDesc: MethodTypeDesc, vararg args: ConstantDesc
    ) {
        with(InvokeDynamicMethodInstruction(
            bootstrapMethod, id, methodTypeDesc, listOf(*args)
        ))
    }

    fun invokeConstructor(
        owner: ClassDesc, constructorTypeDesc: MethodTypeDesc,
        argsInstructions: InstructionsSet.() -> Unit = {}
    ) {
        with(InvokeConstructorInstruction(
            owner, constructorTypeDesc, argsInstructions, false
        ))
    }

    fun invokeSuperClass(
        owner: ClassDesc, constructorTypeDesc: MethodTypeDesc,
        argsInstructions: InstructionsSet.() -> Unit = {}
    ) {
        with(InvokeConstructorInstruction(
            owner, constructorTypeDesc, argsInstructions, true
        ))
    }



    fun loadConstant(constant: Int) = with(LoadConstantInstruction(constant))
    fun loadConstant(constant: Long) = with(LoadConstantInstruction(constant))
    fun loadConstant(constant: Float) = with(LoadConstantInstruction(constant))
    fun loadConstant(constant: Double) = with(LoadConstantInstruction(constant))
    fun loadConstant(constant: Boolean) = with(LoadConstantInstruction(constant))
    fun loadConstant(constant: String) = with(LoadConstantInstruction(constant))
    fun loadConstant(constant: ClassDesc) = with(LoadConstantInstruction(constant))
    fun loadNull() = with(LoadConstantInstruction())

    fun loadThisReference() = with(LoadThisReferenceInstruction())



    fun getLocal(type: ClassDesc, slot: Int) {
        with(GetLocalInstruction(type, slot))
    }

    fun storeLocal(type: ClassDesc, slot: Int) {
        with(StoreLocalInstruction(type, slot))
    }

    fun setLocalName(slot: Int, id: String, type: ClassDesc, startLabelUuid: Uuid, endLabelUuid: Uuid) {
        with(SetLocalNameInstruction(slot, id, type, startLabelUuid, endLabelUuid))
    }



    fun newReferenceArray(type: ClassDesc) {
        with(NewReferenceArrayInstruction(type))
    }

    fun storeReferenceIntoArray() {
        with(StoreReferenceIntoArrayInstruction())
    }



    fun negateNumber(type: NumberType) {
        with(NegateNumberInstruction(type))
    }

    fun convertToNumberType(from: NumberType, to: NumberType) {
        with(ConvertToNumberTypeInstruction(from, to))
    }

    fun convertToBooleanType(isFromBoxed: Boolean, isToBoxed: Boolean) {
        with(ConvertToBooleanTypeInstruction(isFromBoxed, isToBoxed))
    }

    fun arithmeticOperation(type: NumberType, operation: ArithmeticOperation) {
        with(ArithmeticOperationInstruction(type, operation))
    }

    fun logicalOperation(operation: LogicalOperation) {
        with(LogicalOperationInstruction(operation))
    }



    fun instanceOf(target: ClassDesc) {
        with(InstanceOfInstruction(target))
    }

    fun cast(type: ClassDesc) {
        with(CastInstruction(type))
    }

    fun returnValue(classDesc: ClassDesc?) {
        with(ReturnInstruction(classDesc))
    }

    fun returnVoid() {
        returnValue(null)
    }



    fun duplicate() {
        with(DuplicateInstruction())
    }

    fun pop() {
        with(PopInstruction())
    }



    fun createLabel(): Uuid {
        val uuid = Uuid.random()
        bytecodeBuilders.addLabel(uuid)
        return uuid
    }

    fun initLabel(uuid: Uuid) {
        with(InitLabelInstruction(uuid))
    }

    fun createAndInitLabel(): Uuid {
        val uuid = createLabel()
        initLabel(uuid)
        return uuid
    }

    fun bindLabel(uuid: Uuid) {
        with(BindLabelInstruction(uuid))
    }

    fun gotoLabel(uuid: Uuid) {
        with(GotoLabelInstruction(uuid))
    }



    fun gotoLabelIfNonNull(uuid: Uuid) {
        with(GotoLabelIfNonNullInstruction(uuid))
    }

    fun gotoLabelIfEqualsZero(uuid: Uuid) {
        with(GotoLabelIfEqualsZeroInstruction(uuid))
    }

    fun gotoLabelIfNotEqualsZero(uuid: Uuid) {
        with(GotoLabelIfNotEqualsZeroInstruction(uuid))
    }

    fun gotoLabelIfComparisonTrue(type: NumberType, operation: ComparisonOperation, uuid: Uuid) {
        with(GotoLabelIfComparisonTrueInstruction(type, operation, uuid))
    }



    companion object {
        private fun toIntFlags(accessFlags: Collection<AccessFlag>): Int {
            return accessFlags
                .map { it.mask() }
                .reduceOrNull { i1, i2 -> i1 or i2 } ?: 0
        }
    }
}



@Suppress("PLATFORM_CLASS_MAPPED_TO_KOTLIN")
val String.java get() = this as java.lang.String